### PR TITLE
3.next view builder

### PR DESCRIPTION
**Pull request by dereuromark**
- Start fixing up ViewBuilder API. Also fixed a few wrong doc blocks.
- Remove deprecated method usage in tests. Make typehinting possible which reveals coding errors.

### CODE REVIEW

**Comment by dereuromark** [comment]
An idea where these "Trying to clone an uncloneable object of class Mock_Collection" errors come from?
Locally it was working fine.

**Comment by ADmad** [comment]
Why change the old deprecated methods to use the new methods instead if directly modifying the protected properties like they currently do?

**Comment by dereuromark** [comment]
It is a bit less code.
And it feels a bit more DRY.

**Comment by ADmad** [comment]
Changing old methods looks like adding more code to me  plus now you have double the amount of method calls when using the old methods.

Not really a fan of splitting the get/set methods for classes like ViewBuilder where you rarely use the methods as getter.

```
$this->viewBuilder()
    ->setTemplate()
    ->setLayout()
    ->setClassname();
```
Looks repitative and ugly to me. Also for each method i have to type 3 extra characters to get decent autocompletion else one would have to scroll through the list.

**Comment by dereuromark** [comment]
Time issues are irrelevant for those deprecated methods.

**Comment by ADmad** [comment]
> Time issues are irrelevant for those deprecated methods.

It does matter. Not everyone is going to go about updating their code bases as soon as 3.4 is released. Some might not till even 4.0 is released.

> It doesnt matter how ugly it looks to you. 

It matters to me since I am using the framework :)

> The consensus of pretty much all developers is that this is a good way to go, using SOLID and all the advantages of those split methods I hopefully do not have to mention.

There are many things that others follow which CakePHP doesn't.  It's the uniqueness of CakePHP which entices me not the similarity with the rest. While this may be the majority consensus of our dev team too I want to voice my opinion.

> This is the only valid approach for a clean API - and as a sidenote also for proper IDE usage in general (right now it aint really working as you might have noticed). Refs #9701

Not sure what IDE usage you are referring too but I don't use an IDE. I was just referring to basic completing providers by editors.

**Comment by dereuromark** [comment]
Come one, now you are fishing for arguments, that is just being dense.
ViewBuilder is never executed in a loop, not even in the worst case of the world this would increase anything beyond 0.00001 seconds...

You should care for "all" using the framework.

**Comment by thinkingmedia** [comment]
@dereuromark @ADmad 

1. All enhancements for the `3.next` branch should have a corresponding issue opened first.
2. If there is a problem with the change, then the discussion should take place in that issue.
3. Open a pull request after there is a general agreement that it should be done.

If you open a pull request without a discussion first, then you risk trying to justify the work after it's been done. That's when things can get problematic. Save yourself the trouble, and discuss the changes first.

@markstory started refactoring the core for PSR-7 support, which includes changes requiring getter/setter methods. I think there was a misunderstanding that we were making this change on a larger scale in the core. The PSR-7 changes are only related to the networking layer of the core so that the request/response objects could be replaceable (correct me if I'm wrong).

Now, I started another project for `3.next` so that it would include many database changes. Together these changes are making the `3.next` branch a larger upgrade.

What we don't have is a clear roadmap of what the `3.next` objectives are, and I'm not sure that changing method signatures (outside of the network layer) is going to be generally liked.

I also tried to change some interface signatures and was faced with the same reaction from members of the team.

I think the point here is that `3.next` is approaching a point where people are getting worried about all the changes. So I ask that we all work together to communicate better so that we don't waste time and things don't get heated.

I hope this clarifies everything.

**Comment by dereuromark** [comment]
@thinkingmedia I see it quite differently.
For me 4.x should have a clean API again, and thus helping in 3.4 to reach that while at the same time having almost 0 upgrade efforts between 3.x and 4.0 then due to early on API cleanup (with deprecations) is a big win IMO.

File: tests/TestCase/View/HelperTest.php [hunk 1]
```diff
@@ -44,19 +44,6 @@ class TestHelper extends Helper
...
-     * @param string $options
-     * @param string $exclude
-     * @param string $insertBefore
-     * @param string $insertAfter
-     * @return void
-     */
-    public function parseAttributes($options, $exclude = null, $insertBefore = ' ', $insertAfter = null)
-    {
-        return $this->_parseAttributes($options, $exclude, $insertBefore, $insertAfter);
-    }
```
Comment (markstory) [inline tests/TestCase/View/HelperTest.php]: :clap:

**Comment by dereuromark** [comment]
Test fails are unrelated, what shall I do? Sqash and retrigger them?

**Comment by dereuromark** [comment]
Looks like the mock issues went away again after retriggering the builds.

**Comment by dereuromark** [comment]
Can we proceed with this PR?

**Comment by dereuromark** [comment]
@markstory @lorenzo Is there a reason for perfectly fine PRs to become stale for no reason here over weeks of time? I would hate to have to rebase those again a few times if it can be avoided.

**Comment by chinpei215** [comment]
@dereuromark I think you need to remove all deprecated method calls from the core, at the same time of doing this. Otherwise, Cake would warn people to use the deprecated methods [here](https://github.com/cakephp/cakephp/blob/1f53eb67ea27f43d4c0ec3dbdac12cc363922c8d/src/Controller/Controller.php#L335-L343), for example. Also, this change makes it difficult to extend ViewBuilder. How can people do it without overriding the deprecated methods, if they are called in the core?

**Comment by dereuromark** [comment]
This is the 2nd step. This can be done after merging and before releasing beta or RC.

**Comment by dereuromark** [comment]
I would like to start removing the deprecations in the code once things have been merged.

**Comment by lorenzo** [comment]
@dereuromark no specific reason other than lack of time to follow up

**Pull request merged**